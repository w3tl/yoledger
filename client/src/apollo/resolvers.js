export const defaults = {
  settings: {
    __typename: 'AppSettings',
    transaction: {
      __typename: 'TransactionSettings',
      period: 'DAY',
      date: new Date().toISOString(),
    },
  },
};

export function getDateStart(period, date) {
  if (period === 'DAY') return date;
  const day = date.getDay();
  const diff = date.getDate() - day + (day === 0 ? -6 : 1); // adjust when day is sunday
  return new Date(date.setDate(diff));
}

export const Mutation = {
  changeTransactionPagePeriod(_, { period, date }, { cache }) {
    const dateStart = getDateStart(period, new Date(date)).toISOString();
    cache.writeData({
      data: {
        settings: {
          __typename: 'AppSettings',
          transaction: {
            __typename: 'TransactionSettings',
            period,
            date: dateStart,
          },
        },
      },
    });

    return { __typename: 'TransactionSettings', period, date: dateStart };
  },
};
