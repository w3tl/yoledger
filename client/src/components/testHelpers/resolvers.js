import dateResolver from './date';
import assets from '../Asset/mockData';
import { expenses, incomes } from '../Categories/mockData';
import { accounts, budgets as rawBudgets } from '../Budget/mockData';
import transactions from '../Transaction/mockData';

const accountsArray = assets.concat(expenses).concat(incomes);
const budgets = rawBudgets.map(({ date, ...fields }) => ({ date: new Date(date), ...fields }));

export default {
  Date: dateResolver,
  Query: {
    account(root, { id }) {
      if (id === '666') return { name: 'new account' };
      return accountsArray.find(a => a.id === id);
    },
    accounts(root, { type }) {
      return accountsArray.filter(a => a.type === type);
    },
    budgets() {
      return { accounts };
    },
    budget(root, { account, dateStart, dateEnd }) {
      return budgets // eslint-disable-next-line max-len
        .filter(b => (
          b.account.name === account
          && b.date.getTime() >= dateStart.getTime()
          && b.date.getTime() <= new Date(dateEnd).getTime()));
    },
    transactions() {
      return transactions.map(({ date, ...other }) => ({ date: new Date(date), ...other }));
    },
    transaction(root, { id }) {
      if (id === '666') return { name: 'new account' };
      const { date, ...other } = transactions.find(a => a.id === id);
      return { date: new Date(date), ...other };
    },
  },
  Mutation: {
    upsertBudget(root, { input: { account, date, amount } }) {
      const existingBudget = budgets
        .find(b => (
          b.account.name === account.name && b.date.getTime() === new Date(date).getTime()
        ));
      const acc = accounts.find(a => a.name === account) || {
        id: '687',
        name: account,
      };
      return {
        success: true,
        budget: {
          id: '99',
          ...existingBudget,
          account: acc,
          date,
          amount,
        },
      };
    },
    addAccount(root, { input }) {
      return {
        account: { id: '666', ...input },
      };
    },
    addTransaction(root, { input }) {
      return {
        transaction: {
          id: '666',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    deleteTransaction(root, { id }) {
      return {
        success: true,
        id,
      };
    },
    updateTransaction(root, { id, input }) {
      return {
        transaction: {
          ...transactions.find(t => t.id === id),
          id: '888',
          ...input,
          source: accountsArray.find(a => a.name === input.source),
          destination: accountsArray.find(a => a.name === input.destination),
        },
      };
    },
    signin(root, { login }) {
      return {
        token: `{\\"userId\\":\\"${login}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
    signout() {
      return {
        success: true,
      };
    },
    changePassword() {
      return {
        token: `{\\"userId\\":\\"${'admin'}\\"};secret;{\\"expiresIn\\":10}`,
      };
    },
  },
};
