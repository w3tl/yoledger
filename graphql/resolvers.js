import {
  Query as UserQuery,
  User,
} from './user';
import {
  Query as AccountQuery,
  Mutation as AccountMutation,
  Account,
} from './account';
import {
  Query as TransactionQuery,
  Mutation as TransactionMutation,
  Transaction,
} from './transaction';
import {
  Query as BudgetQuery,
  Mutation as BudgetMutation,
  Budget,
  BudgetPlan,
} from './budget';
import dateResolver from './date';

export default {
  Query: Object.assign({}, UserQuery, AccountQuery, TransactionQuery, BudgetQuery),
  Mutation: Object.assign({}, AccountMutation, TransactionMutation, BudgetMutation),
  User,
  Account,
  Transaction,
  Budget,
  BudgetPlan,
  Date: dateResolver,
};
