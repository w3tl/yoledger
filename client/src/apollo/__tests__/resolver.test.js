import { getDateStart, Mutation } from '../resolvers';

describe('resolver', () => {
  it('getDateStart should return correct date for any period', () => {
    const dateForDay = getDateStart('DAY', new Date('2018-01-01'));
    expect(dateForDay.toISOString()).toMatchSnapshot('DAY');
    const dateForWeek = getDateStart('WEEK', new Date('2018-01-01'));
    expect(dateForWeek.toISOString()).toMatchSnapshot('start WEEK in 2018-01-01');
    const dateForWeek2 = getDateStart('WEEK', new Date('2018-01-11'));
    expect(dateForWeek2.toISOString()).toMatchSnapshot('start WEEK in 2018-01-11');
    const dateForWeek3 = getDateStart('WEEK', new Date('2018-06-13'));
    expect(dateForWeek3.toISOString()).toMatchSnapshot('start WEEK in 2018-06-13');
  });

  describe('Mutation changeTransactionPagePeriod', () => {
    const date = '2018-01-01';
    const cache = {
      writeData: jest.fn(),
    };
    it('should write to cache and return', () => {
      const result = Mutation.changeTransactionPagePeriod(null, { period: 'WEEK', date }, { cache });
      expect(cache.writeData.mock.calls[0]).toMatchSnapshot('writeData');
      expect(result).toMatchSnapshot('WEEK 2018-01-01');
    });
  });
});
