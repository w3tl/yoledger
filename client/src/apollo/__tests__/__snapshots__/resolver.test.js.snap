// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DAY 1`] = `"2018-01-01T00:00:00.000Z"`;

exports[`WEEK 2018-01-01 1`] = `
Object {
  "__typename": "TransactionSettings",
  "date": "2018-01-01T00:00:00.000Z",
  "period": "WEEK",
}
`;

exports[`start WEEK in 2018-01-01 1`] = `"2018-01-01T00:00:00.000Z"`;

exports[`start WEEK in 2018-01-11 1`] = `"2018-01-08T00:00:00.000Z"`;

exports[`start WEEK in 2018-06-13 1`] = `"2018-06-11T00:00:00.000Z"`;

exports[`writeData 1`] = `
Array [
  Object {
    "data": Object {
      "settings": Object {
        "__typename": "AppSettings",
        "transaction": Object {
          "__typename": "TransactionSettings",
          "date": "2018-01-01T00:00:00.000Z",
          "period": "WEEK",
        },
      },
    },
  },
]
`;
